widget/entry.go

```diff
@@ -849,47 +862,51 @@ func (e *Entry) textWrap() fyne.TextWrap {
 }
 
 func (e *Entry) updateMousePointer(ev *fyne.PointEvent, rightClick bool) {
-	if !e.focused && !e.Disabled() {
+	if !e.Focused() && !e.Disabled() {
 		e.FocusGained()
 	}
 
-	rowHeight := e.textProvider().charMinSize().Height
-	row := int(math.Floor(float64(ev.Position.Y-theme.Padding()) / float64(rowHeight)))
-	col := 0
-	if row < 0 {
-		row = 0
-	} else if row >= e.textProvider().rows() {
-		row = e.textProvider().rows() - 1
-		col = 0
-	} else {
-		col = e.cursorColAt(e.textProvider().row(row), ev.Position)
-	}
+	e.SetFieldsAndRefresh(func() {
 
-	e.Lock()
-	if !rightClick || rightClick && !e.selecting {
-		e.CursorRow = row
-		e.CursorColumn = col
-	}
+		rowHeight := e.textProvider().charMinSize().Height
+		row := int(math.Floor(float64(ev.Position.Y-theme.Padding()) / float64(rowHeight)))
+		col := 0
+		if row < 0 {
+			row = 0
+		} else if row >= e.textProvider().rows() {
+			row = e.textProvider().rows() - 1
+			col = 0
+		} else {
+			col = e.cursorColAt(e.textProvider().row(row), ev.Position)
+		}
 
-	if !e.selecting {
-		e.selectRow = row
-		e.selectColumn = col
-	}
-	e.Unlock()
-	e.Refresh()
+		if !rightClick || rightClick && !e.selecting {
+			e.CursorRow = row
+			e.CursorColumn = col
+		}
+
+		if !e.selecting {
+			e.selectRow = row
+			e.selectColumn = col
+		}
+	})
 }
 
 // updateText updates the internal text to the given value
 func (e *Entry) updateText(text string) {
-	changed := e.Text != text
-	e.Lock()
-	e.Text = text
-	e.Unlock()
-	if changed && e.OnChanged != nil {
-		e.OnChanged(text)
-	}
+	var callback func(string)
+	e.SetFieldsAndRefresh(func() {
+		changed := e.Text != text
+		e.Text = text
 
-	e.Refresh()
+		if changed && e.OnChanged != nil {
```

Comment:
Nil check not needed here, it is already checked at 907
```suggestion
		if changed {
```